Windows 10 Pro HVM does not work with Mirage Firewall

Hi Thomas,

First of all, not being much technical may prevent me from realizing all you've done to create this qubes alternative firewall but from what I've read about Mirage unikernel and tried with v0.5, I can say it's a really great advance for the Qubes community. I hope your work will be incorporated in future Qubes releases as an alternative to sys-firewall.

I have a Mirage Firewall VM v.05 that works with HVM (Linux Mint) and
Debian/Fedora template-based PVMs.

My Windows 10 HVM works with sys-firewall using auto DHCP, while on Linux Mint I have to setup IP, mask and gateway (works with both sys-firewall and mirage-firewall as well).

On Windows Mirage did not work and I've tried setting up networking manually by adding its IP, mask and gateway and rebooting but it did not work.

I've followed all the ideas from here (https://www.windowscentral.com/how-regain-internet-access-after-installing-update-windows-10) and it still did not work.

One last piece of information, my Windows 10 Pro was successfully activated using a key I provided.

Any ideas? This is not critical, since I can continue using sys-firewall, but would love to free some memory by using Mirage.

Below I've added logs from guest-mirage-firewall.log. My Windows VM is 10.137.0.21.

Best Regards,

Claudio Chinicz

2019-04-18 11:20:10 -00:00: INF [client_net] Client 18 (IP: 10.137.0.21) ready
2019-04-18 11:20:10 -00:00: INF [ethernet] Connected Ethernet interface 00:16:3e:5e:6c:00
2019-04-18 11:20:11 -00:00: INF [client_net] add client vif {domid=17;device_id=0}
2019-04-18 11:20:11 -00:00: INF [qubes.db] got rm "/qubes-iptables-domainrules/"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables-header" = "# Generated by Qubes Core on Thu Apr 18 14:20:11 2019\n*filter\n:INPUT DROP [0:0]\n:FORWARD DROP [0:0]\n:OUTPUT ACCEPT [0:0]\n-A INPUT -i vif+ -p udp -m udp --dport 68 -j DROP\n-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT\n-A INPUT -p icmp -j ACCEPT\n-A INPUT -i lo -j ACCEPT\n-A INPUT -j REJECT --reject-with icmp-host-prohibited\n-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT\n-A FORWARD -i vif+ -o vif+ -j DROP\nCOMMIT\n"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables-domainrules/10" = "*filter\n-A FORWARD -s 10.137.0.18 -j ACCEPT\n-A FORWARD -s 10.137.0.18 -j DROP\nCOMMIT\n"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables-domainrules/17" = "*filter\n-A FORWARD -s 10.137.0.21 -j ACCEPT\n-A FORWARD -s 10.137.0.21 -j DROP\nCOMMIT\n"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables-domainrules/14" = "*filter\n-A FORWARD -s 10.137.0.13 -j ACCEPT\n-A FORWARD -s 10.137.0.13 -j DROP\nCOMMIT\n"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables-domainrules/9" = "*filter\n-A FORWARD -s 10.137.0.8 -j ACCEPT\n-A FORWARD -s 10.137.0.8 -j DROP\nCOMMIT\n"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-iptables" = "reload"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/mapped-ip/10.137.0.21/visible-ip" = "10.137.0.21"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/mapped-ip/10.137.0.21/visible-gateway" = "10.137.0.23"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21/0000" = "action=accept"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21/policy" = "drop"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21" = ""
2019-04-18 11:20:11 -00:00: INF [qubes.db] got rm "/qubes-firewall/10.137.0.21/"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21/0000" = "action=accept"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21/policy" = "drop"
2019-04-18 11:20:11 -00:00: INF [qubes.db] got update: "/qubes-firewall/10.137.0.21" = ""
2019-04-18 11:20:22 -00:00: INF [client_eth] who-has 10.137.0.1?
2019-04-18 11:20:22 -00:00: INF [client_eth] unknown address; not responding
2019-04-18 11:20:22 -00:00: WRN [client_net] Incorrect source IP 0.0.0.0 in IP packet from 10.137.0.21 (dropping)
2019-04-18 11:20:22 -00:00: WRN [client_net] Incorrect source IP 10.137.0.1 in IP packet from 10.137.0.21 (dropping)
2019-04-18 11:20:22 -00:00: INF [client_eth] who-has 10.137.0.1?
2019-04-18 11:20:22 -00:00: INF [client_eth] unknown address; not responding
2019-04-18 11:20:22 -00:00: INF [client_eth] who-has 10.137.0.21?
2019-04-18 11:20:22 -00:00: INF [client_eth] ignoring request for client's own IP
2019-04-18 11:20:22 -00:00: WRN [client_net] Ignored unknown IPv4 message: Ignoring non-TCP/UDP packet: IPv4 packet 10.137.0.21 -> 224.0.0.22: id 46e6, off 0 proto 2, ttl 1, options
94 04 00 00
2019-04-18 11:20:22 -00:00: INF [client_eth] who-has 10.137.0.21?
2019-04-18 11:20:22 -00:00: INF [client_eth] ignoring request for client's own IP
2019-04-18 11:20:22 -00:00: INF [client_eth] who-has 10.137.0.1?
2019-04-18 11:20:22 -00:00: INF [client_eth] unknown address; not responding
2019-04-18 11:20:22 -00:00: WRN [client_net] Ignored unknown IPv4 message: Ignoring non-TCP/UDP packet: IPv4 packet 10.137.0.21 -> 224.0.0.22: id 46e7, off 0 proto 2, ttl 1, options
94 04 00 00
2019-04-18 11:20:22 -00:00: WRN [client_net] Ignored unknown IPv4 message: Ignoring non-TCP/UDP packet: IPv4 packet 10.137.0.21 -> 224.0.0.22: id 46e8, off 0 proto 2, ttl 1, options
94 04 00 00
2019-04-18 11:20:22 -00:00: WRN [client_net] Incorrect source IP 0.0.0.0 in IP packet from 10.137.0.21 (dropping)
2019-04-18 11:20:22 -00:00: WRN [client_net] Incorrect source IP 10.137.0.1 in IP packet from 10.137.0.21 (dropping)
2019-04-18 11:20:22 -00:00: WRN [client_net] Ignored unknown IPv4 message: Ignoring non-TCP/UDP packet: IPv4 packet 10.137.0.21 -> 224.0.0.22: id 46e9, off 0 proto 2, ttl 1, options
94 04 00 00
2019-04-18 11:20:22 -00:00: WRN [firewall] Failed to add NAT rewrite rule: Cannot NAT this packet (IPv4 packet 10.137.0.21 -> 224.0.0.251: id e7de, off 0 proto 17, ttl 1, options
 UDP port 5353 -> 5353)
2019-04-18 11:20:22 -00:00: WRN [firewall] Failed to add NAT rewrite rule: Cannot NAT this packet (IPv4 packet 10.137.0.21 -> 224.0.0.252: id 211e, off 0 proto 17, ttl 1, options
 UDP port 53180 -> 5355)
2019-04-18 11:20:22 -00:00: WRN [firewall] Failed to add NAT rewrite rule: Cannot NAT this packet (IPv4 packet 10.137.0.21 -> 224.0.0.251: id e7df, off 0 proto 17, ttl 1, options

Hi Thomas,

I followed the instructions but there were errors and warnings that I extracted from the console (I've skipped most of the lines and included just the errors/warnings and some lines close to them to help you figure out where they happened). Please see below.

One question, it's written "Note: the object files are stored in the _build directory....". I could not find this _build directory, maybe because of the errors. Should it be under /docker or /qubes-mirage-firewall?

Thanks,
Claudio

Get:15 http://deb.debian.org/debian stretch/main amd64 xml-core all 0.17 [23.2 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 16.1 MB in 2s (6060 kB/s)

[ERROR] The compilation of xenstore failed at "/home/opam/.opam/4.07/bin/dune build -p xenstore -j 71".
[ERROR] The compilation of shared-memory-ring failed at "/home/opam/.opam/4.07/bin/dune build -p shared-memory-ring -j 71".
[ERROR] The compilation of xen-gnt failed at "/home/opam/.opam/4.07/bin/dune build -p xen-gnt -j 71".

#=== ERROR while installing mirage-stack.1.4.0 ================================#
# path                 ~/.opam/4.07/.opam-switch/build/mirage-stack.1.4.0
# command              /usr/bin/install -m 0644 /home/opam/.opam/4.07/.opam-switch/build/mirage-stack.1.4.0/_build/install/default/lib/mirage-stack/mirage_stack.cmt /home/opam/.opam/4.07/lib/mirage-stack/mirage_stack.cmt
# exit-code            1
# env-file             ~/.opam/log/log-8-05304b.env
# output-file          ~/.opam/log/log-8-05304b.out


#=== ERROR while installing mirage-protocols-lwt.2.0.0 ========================#
"install": command not found.

#=== ERROR while installing functoria.2.2.3 ===================================#
# path                 ~/.opam/4.07/.opam-switch/build/functoria.2.2.3
# command              /usr/bin/install -m 0644 /home/opam/.opam/4.07/.opam-switch/build/functoria.2.2.3/_build/install/default/lib/functoria/functoria.cmti /home/opam/.opam/4.07/lib/functoria/functoria.cmti
# exit-code            1
# env-file             ~/.opam/log/log-8-01ae51.env
# output-file          ~/.opam/log/log-8-01ae51.out


#=== ERROR while compiling xenstore.2.1.0 =====================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/xenstore.2.1.0
# command              ~/.opam/4.07/bin/dune build -p xenstore -j 71
# exit-code            1
# env-file             ~/.opam/log/xenstore-8-ddda1d.env
# output-file          ~/.opam/log/xenstore-8-ddda1d.out
### output ###
#     ocamlopt .ppx/fae5e047b89b86eceafabfbc9ce59292/ppx.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlopt.opt -o .ppx/fae5e047b89b86eceafabfbc9ce59292/ppx.exe -I /home/opam/.opam/4.07/lib/base/caml -I /home/opam/.opam/4.07/lib/ocaml-migrate-parsetree -I /home/opam/.opam/4.07/lib/ocaml/compiler-libs -I /home/opam/.opam/4.07/lib/parsexp -I /home/opam/.opam/4.07/lib/ppx_cstruct -I /home/opam/.opam/4.07/lib/ppx_derivers -I /home/opam/.opam/4.07[...]
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking


#=== ERROR while compiling xen-gnt.4.0.0 ======================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/xen-gnt.4.0.0
# command              ~/.opam/4.07/bin/dune build -p xen-gnt -j 71
# exit-code            2
# env-file             ~/.opam/log/xen-gnt-8-f736a3.env
# output-file          ~/.opam/log/xen-gnt-8-f736a3.out

[cmdliner.1.0.3] synchronised from git+https://github.com/talex5/cmdliner.git#repro-builds
[WARNING] Failed checks on cmdliner package definition from source at git+https://github.com/talex5/cmdliner.git#repro-builds:
             error 57: Synopsis and description must not be both empty
cmdliner is now pinned to git+https://github.com/talex5/cmdliner.git#repro-builds (version 1.0.3)
 ---> 2674c4075db6


#=== ERROR while compiling shared-memory-ring.3.1.0 ===========================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.07.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.07/.opam-switch/build/shared-memory-ring.3.1.0
# command              ~/.opam/4.07/bin/dune build -p shared-memory-ring -j 71
# exit-code            2
# env-file             ~/.opam/log/shared-memory-ring-8-9acf8d.env
# output-file          ~/.opam/log/shared-memory-ring-8-9acf8d.out
### output ###
# Error: Assembler error, input left in file /tmp/camlstartup5f01c1.s
# [...]
# (cd _build/default && /home/opam/.opam/4.07/bin/ocamlmklib.opt -g -o lib/shared_memory_ring_stubs lib/barrier_stubs.o)
# collect2: error: ld returned 1 exit status
# Fatal error: exception Sys_error("No space left on device")
# Raised by primitive operation at file "stdlib.ml", line 377, characters 19-27
# Called from file "src/stdune/exn.ml", line 30, characters 19-28
# Called from file "list.ml", line 106, characters 12-15
# Called from file "src/hooks.ml", line 22, characters 4-50
# Called from file "stdlib.ml", line 545, characters 62-65
# Called from file "stdlib.ml" (inlined), line 548, characters 20-39
# Called from file "stdlib.ml", line 551, characters 2-15
# Called from file "bin/main_dune.ml", line 1, characters 9-21


The former state can be restored with:
    opam switch import "/home/opam/.opam/4.07/.opam-switch/backup/state-20190501133041.export"

--- a/Dockerfile
+++ b/Dockerfile
@@ -7,7 +7,7 @@ FROM ocaml/opam2@sha256:f7125924dd6632099ff98b2505536fe5f5c36bf0beb24779431bb62b
 # Pin last known-good version for reproducible builds.
 # Remove this line (and the base image pin above) if you want to test with the
 # latest versions.
-RUN git fetch origin && git reset --hard e77756e92274790668ed1f6f998d66fa2e744fb6 && opam update
+RUN git fetch origin && git reset --hard d1b2a1cbc28d43926b37e61f46fc403b48ab9c23 && opam update
 
 RUN sudo apt-get install -y m4 libxen-dev pkg-config
 RUN opam pin add -yn cmdliner 'https://github.com/talex5/cmdliner.git#repro-builds'

--- a/build-with-docker.sh
+++ b/build-with-docker.sh
@@ -5,5 +5,5 @@ docker build -t qubes-mirage-firewall .
 echo Building Firewall...
 docker run --rm -i -v `pwd`:/home/opam/qubes-mirage-firewall qubes-mirage-firewall
 echo "SHA2 of build:   $(sha256sum qubes_firewall.xen)"
-echo "SHA2 last known: 765cf16c2e85feb7e5dfd3e409a3013c91c2b07f5680ed9f4e487e27213f1355"
+echo "SHA2 last known: dbf7460fa628bea5d132a96fe7ba2cd832e3d9da7005ae74f6a124957f4848ea"
 echo "(hashes should match for released versions)"

--- a/client_eth.ml
+++ b/client_eth.ml
@@ -27,16 +27,16 @@ let client_gw t = t.client_gw
 let add_client t iface =
   let ip = iface#other_ip in
   let rec aux () =
-    if IpMap.mem ip t.iface_of_ip then (
+    match IpMap.find ip t.iface_of_ip with
+    | Some old ->
       (* Wait for old client to disappear before adding one with the same IP address.
          Otherwise, its [remove_client] call will remove the new client instead. *)
-      Log.info (fun f -> f "Waiting for old client %a to go away before accepting new one" Ipaddr.V4.pp ip);
+      Log.info (fun f -> f ~header:iface#log_header "Waiting for old client %s to go away before accepting new one" old#log_header);
       Lwt_condition.wait t.changed >>= aux
-    ) else (
+    | None ->
       t.iface_of_ip <- t.iface_of_ip |> IpMap.add ip iface;
       Lwt_condition.broadcast t.changed ();
       Lwt.return_unit
-    )
   in
   aux ()
 
@@ -83,16 +83,18 @@ module ARP = struct
 
   let input_query t arp =
     let req_ipv4 = arp.Arp_packet.target_ip in
-    Log.info (fun f -> f "who-has %s?" (Ipaddr.V4.to_string req_ipv4));
+    let pf (f : ?header:string -> ?tags:_ -> _) fmt =
+      f ~header:t.client_link#log_header ("who-has %a? " ^^ fmt) Ipaddr.V4.pp req_ipv4
+    in
     if req_ipv4 = t.client_link#other_ip then (
-      Log.info (fun f -> f "ignoring request for client's own IP");
+      Log.info (fun f -> pf f "ignoring request for client's own IP");
       None
     ) else match lookup t req_ipv4 with
       | None ->
-        Log.info (fun f -> f "unknown address; not responding");
+        Log.info (fun f -> pf f "unknown address; not responding");
         None
       | Some req_mac ->
-        Log.info (fun f -> f "responding to: who-has %s?" (Ipaddr.V4.to_string req_ipv4));
+        Log.info (fun f -> pf f "responding with %a" Macaddr.pp req_mac);
         Some { Arp_packet.
                operation = Arp_packet.Reply;
                (* The Target Hardware Address and IP are copied from the request *)
@@ -105,15 +107,16 @@ module ARP = struct
   let input_gratuitous t arp =
     let source_ip = arp.Arp_packet.source_ip in
     let source_mac = arp.Arp_packet.source_mac in
+    let header = t.client_link#log_header in
     match lookup t source_ip with
     | Some real_mac when Macaddr.compare source_mac real_mac = 0 ->
-      Log.info (fun f -> f "client suggests updating %s -> %s (as expected)"
+      Log.info (fun f -> f ~header "client suggests updating %s -> %s (as expected)"
                    (Ipaddr.V4.to_string source_ip) (Macaddr.to_string source_mac));
     | Some other_mac ->
-      Log.warn (fun f -> f "client suggests incorrect update %s -> %s (should be %s)"
+      Log.warn (fun f -> f ~header "client suggests incorrect update %s -> %s (should be %s)"
                    (Ipaddr.V4.to_string source_ip) (Macaddr.to_string source_mac) (Macaddr.to_string other_mac));
     | None ->
-      Log.warn (fun f -> f "client suggests incorrect update %s -> %s (unexpected IP)"
+      Log.warn (fun f -> f ~header "client suggests incorrect update %s -> %s (unexpected IP)"
                    (Ipaddr.V4.to_string source_ip) (Macaddr.to_string source_mac))
 
   let input t arp =

--- a/client_eth.ml
+++ b/client_eth.ml
@@ -70,7 +70,11 @@ module ARP = struct
 
   let lookup t ip =
     if ip = t.net.client_gw then Some t.client_link#my_mac
-    else None
+    else if (Ipaddr.V4.to_bytes ip).[3] = '\x01' then (
+      Log.info (fun f -> f ~header:t.client_link#log_header
+                   "Request for %a is invalid, but pretending it's me (see Qubes issue #5022)" Ipaddr.V4.pp ip);
+      Some t.client_link#my_mac
+    ) else None
   (* We're now treating client networks as point-to-point links,
      so we no longer respond on behalf of other clients. *)
     (*

--- a/client_net.ml
+++ b/client_net.ml
@@ -26,17 +26,20 @@ let writev eth dst proto fillfn =
        Lwt.return ()
     )
 
-class client_iface eth ~gateway_ip ~client_ip client_mac : client_link = object
-  val queue = FrameQ.create (Ipaddr.V4.to_string client_ip)
-  method my_mac = ClientEth.mac eth
-  method other_mac = client_mac
-  method my_ip = gateway_ip
-  method other_ip = client_ip
-  method writev proto fillfn =
-    FrameQ.send queue (fun () ->
-        writev eth client_mac proto fillfn
-    )
-end
+class client_iface eth ~domid ~gateway_ip ~client_ip client_mac : client_link =
+  let log_header = Fmt.strf "dom%d:%a" domid Ipaddr.V4.pp client_ip in
+  object
+    val queue = FrameQ.create (Ipaddr.V4.to_string client_ip)
+    method my_mac = ClientEth.mac eth
+    method other_mac = client_mac
+    method my_ip = gateway_ip
+    method other_ip = client_ip
+    method writev proto fillfn =
+      FrameQ.send queue (fun () ->
+          writev eth client_mac proto fillfn
+        )
+    method log_header = log_header
+  end
 
 let clients : Cleanup.t Dao.VifMap.t ref = ref Dao.VifMap.empty
 
@@ -76,7 +79,7 @@ let add_vif { Dao.ClientVif.domid; device_id } ~client_ip ~router ~cleanup_tasks
   let client_mac = Netback.frontend_mac backend in
   let client_eth = router.Router.client_eth in
   let gateway_ip = Client_eth.client_gw client_eth in
-  let iface = new client_iface eth ~gateway_ip ~client_ip client_mac in
+  let iface = new client_iface eth ~domid ~gateway_ip ~client_ip client_mac in
   Router.add_client router iface >>= fun () ->
   Cleanup.on_cleanup cleanup_tasks (fun () -> Router.remove_client router iface);
   let fixed_arp = Client_eth.ARP.create ~net:client_eth iface in
@@ -99,7 +102,7 @@ let add_vif { Dao.ClientVif.domid; device_id } ~client_ip ~router ~cleanup_tasks
 (** A new client VM has been found in XenStore. Find its interface and connect to it. *)
 let add_client ~router vif client_ip =
   let cleanup_tasks = Cleanup.create () in
-  Log.info (fun f -> f "add client vif %a" Dao.ClientVif.pp vif);
+  Log.info (fun f -> f "add client vif %a with IP %a" Dao.ClientVif.pp vif Ipaddr.V4.pp client_ip);
   Lwt.async (fun () ->
       Lwt.catch (fun () ->
           add_vif vif ~client_ip ~router ~cleanup_tasks

--- a/client_net.ml
+++ b/client_net.ml
@@ -73,7 +73,7 @@ let add_vif { Dao.ClientVif.domid; device_id } ~client_ip ~router ~cleanup_tasks
   Netback.make ~domid ~device_id >>= fun backend ->
   Log.info (fun f -> f "Client %d (IP: %s) ready" domid (Ipaddr.V4.to_string client_ip));
   ClientEth.connect backend >>= fun eth ->
-  let client_mac = Netback.mac backend in
+  let client_mac = Netback.frontend_mac backend in
   let client_eth = router.Router.client_eth in
   let gateway_ip = Client_eth.client_gw client_eth in
   let iface = new client_iface eth ~gateway_ip ~client_ip client_mac in

--- a/config.ml
+++ b/config.ml
@@ -27,7 +27,7 @@ let main =
       package "ethernet";
       package "mirage-protocols";
       package "shared-memory-ring" ~min:"3.0.0";
-      package "netchannel" ~min:"1.10.2";
+      package "netchannel" ~min:"1.11.0" ~pin:"git+https://github.com/mirage/mirage-net-xen.git";
       package "mirage-net-xen";
       package "ipaddr" ~min:"3.0.0";
       package "mirage-qubes";

--- a/fw_utils.ml
+++ b/fw_utils.ml
@@ -30,6 +30,7 @@ end
 class type client_link = object
   inherit interface
   method other_mac : Macaddr.t
+  method log_header : string  (* For log messages *)
 end
 
 (** An Ethernet header from [src]'s MAC address to [dst]'s with an IPv4 payload. *)

